Optional mTLS client certificate metadata verification

peanball opened the pull request:
<!-- Thanks for contributing to 'gorouter'. To speed up the process of reviewing your pull request please provide us with: -->

* A short explanation of the proposed change:

We want to add optional mTLS client certificate metadata verification in addition to the issuer CA validity checks.

Given a particular signing CA (identified by its subject) we want to limit the allowed client certificate subjects that are allowed to send requests.

While the underlying issue is better solved with changes to the PKI and trust relationship between CAs, the PKI or its use by other sub-entities is not always under the control of the operator.

* An explanation of the use cases your change solves

The use case is an upstream proxy managed by a different entity that uses a widely spread private CA for issuing its certificates.

We want to avoid that every user of that CA is able to issue requests to Gorouter, while still explicitly allowing the upstream proxy and its CA certificates, which are short-lived and need rotation.

In the specific use case, we have a CA that controls, which certificates are issued for specific sub-entities, which are clearly identified in the certificate's subject DN. 

* Instructions to functionally test the behavior change using operator interfaces (BOSH manifest, logs, curl, and metrics)

e.g. configuration in routing-release:
```yaml
router:
  verify_client_certificate_metadata:
    - issuer_in_chain:
        common_name: Some Root CA
        country:
          - DE
      valid_cert_subjects:
        - common_name: Awesome Service 1
          organizational_unit:
            - sub-entity4
          organization:
            - Acme, Inc.
          country:
            - US
        - common_name: Awesome Service 2
          organizational_unit:
            - sub-entity7
          organization:
            - Acme, Inc.
          country:
            - US
```

1. Send a request using a client certificate that is signed by a CA that is configured in `client_ca_certs:` 
   but not in the explicit allowed subjects, configured in `verify_client_certificate_metadata:`:
   ```shell
   curl --cacert other-department-5.pem https://app.cf.example.com
   ```
2. Send a request using a client certificate that is signed by a CA that is configured in `client_ca_certs:`
   and in the allowed subjects, configured in `verify_client_certificate_metadata:`:
   ```shell
   curl --cacert awesome-service-2.pem https://app.cf.example.com
   ```

* Expected result after the change

1. The request fails with a TLS handshake error. Gorouter logs the certificate subject and the fact that this certificate was not allowed.
2. succeeds, as desired.

* Current result before the change

1. The request passes through without TLS handshake errors, allows access to a different party that should not have access.
2. succeeds, as desired.


* Links to any other associated PRs

* [x] I have viewed signed and have submitted the Contributor License Agreement

* [x] I have made this pull request to the `main` branch

* [x] I have [run all the unit tests](https://github.com/cloudfoundry/routing-release#running-unit-and-integration-tests).

* [ ] (Optional) I have run Routing Acceptance Tests and Routing Smoke Tests

* [ ] (Optional) I have run CF Acceptance Tests

ameowlia commented on config/config.go:
```diff
@@ -198,6 +203,130 @@ type HTTPRewriteResponses struct {
+// InitClientCertMetadataRules compares the defined rules against client CAs set in `client_ca_certs`. When a rule
+// is found that does not have a corresponding client CA (based on the CA's subject) that matches the rule, startup will fail.
+//
+// This is to avoid defining a rule with a minor typo that would then not apply at all and would make the whole
```
This makes sense that we would want to check for typos, but I worry that failing at runtime is inconvenient and late in the process. Is there a way to fail at templating time instead (or in addition)?

peanball replied:
Yes, at least in addition. While it's unlikely that e.g. Gorouter would be used outside the routing-release, I would prefer to have a safeguard in the Gorouter code anyway.

ameowlia replied:
in addition is great!

ameowlia commented on config/config.go:
```diff
@@ -198,6 +203,130 @@ type HTTPRewriteResponses struct {
 	RemoveHeaders          []HeaderNameValue `yaml:"remove_headers,omitempty"`
 }
 
+// VerifyClientCertificateMetadata defines verification rules for client certificates, which allow additional checks
```
We don't typically have such verbose comments in runtime code, but given the development on this codebase is slowing down and is relatively stable, I think it is helpful and I like it.

ameowlia commented on config/config.go:
```diff
@@ -688,6 +819,18 @@ func (c *Config) buildClientCertPool() error {
 	}
 	c.ClientCAPool = certPool
+
+	if c.VerifyClientCertificateMetadata != nil {
```
Instead of seeing if this property is nil, can you add a separate boolean property `VerifyClientCertificatesBasedOnProvidedMetadata` or something? I have run into issues in the past what I try to turn an obj for a feature into a boolean identifier, because then invariably, we add another property for that feature, and then things become a muddled mess.

ameowlia commented on config/config_test.go:
```diff
@@ -1408,7 +1408,25 @@ route_services_secret_decrypt_only: 1PfbARmvIn6cgyKorA1rqR2d34rBOo+z3qJGz17pi8Y=
+							},
+						}
+					})
+					It("fails to validate", func() {
```
```suggestion
					It("returns a meaningful error", func() {
```

ameowlia commented on router/router.go:
```diff
@@ -274,6 +279,17 @@ func (r *Router) serveHTTPS(server *http.Server, errChan chan error) error {
+	if chains != nil {
+		return config.VerifyClientCertMetadata(r.config.VerifyClientCertificateMetadata, chains, r.logger)
+	}
+	fmt.Printf("Chains length %d", len(chains))
```
```suggestion
```

👻 You probably didn't mean to leave this in :)

ameowlia commented on router/router_test.go:
```diff
@@ -1862,6 +1862,64 @@ var _ = Describe("Router", func() {
+									}
+								})
+								It("successfully serves SSL traffic", func() {
+
```
```suggestion
```

ameowlia commented on config/config.go:
```diff
@@ -198,6 +203,130 @@ type HTTPRewriteResponses struct {
+			return fmt.Errorf("subject not in the list of allowed subjects for CA Subject %q: %q", rule.CASubject, subject)
+		}
+	}
+	return fmt.Errorf("incomplete chain")
```
1. Is there a test for this case?
2. Can you provide a more verbose error message? I don't think I would understand this message without some code digging to find your helpful comments.

peanball replied:
No there is no test yet.

Yes we will add a test and a more verbose error message. It is possible/likely that this will not get through the regular client certificate check already. In that case I would remove it.

peanball replied:
I have provided a more verbose error message and a comment.

I couldn't find a way to get to that point of having an incomplete chain processed by the callback for verifying the peer cert.

ameowlia commented on router/router_test.go:
```diff
@@ -1862,6 +1862,64 @@ var _ = Describe("Router", func() {
 						})
+
+						Context("when verify_client_certificate_metadata is provided ", func() {
+							Context("when the client cert subject is one of the valid subjects for the specified CA subject", func() {
```
What about a case where the cert has more info than is specified in the metadata?

For example...
1. The metadata says the country must be USA. 
1. But the cert includes country is USA, and company is Neopets. 

Hopefully it will succeed because the cert matches all of the restrictions explicitly specified in the metadata?

Can we test a case like this?

peanball replied:
That would actually not match. Internally we convert all of this into a [`pkix.Name`](https://pkg.go.dev/crypto/x509/pkix#Name) as you've likely seen.

The design is for an exact match at the moment, i.e. your configuration needs to be exhaustive.

There is one caveat regarding the matching of `pkix.Name`: We only use the base properties, not `Names` or `ExtraNames` at the moment.

But we can have a test nonetheless that would show exactly that, i.e. that it would not match. Any field that is not defined _must_ be empty right now.

peanball commented:
@ameowlia thanks so much for this thoughtful and thorough review. We'll address the points. Fully agree on having a check in the spec ERB template / at template generation time.

peanball commented:
@ameowlia Could you have another look?

I've updated this PR addressing all comments and provided the companion PR for routing release: https://github.com/cloudfoundry/routing-release/pull/355

b1tamara commented on config/config.go:
```diff
@@ -198,6 +203,132 @@ type HTTPRewriteResponses struct {
+// * the chain does not contain any client certificates (i.e. IsCA == false).
+func checkClientCertificateMetadataRule(chain []*x509.Certificate, logger logger.Logger, rule VerifyClientCertificateMetadataRule) error {
+	for _, cert := range chain {
+		if !cert.IsCA {
+			subject := cert.Subject
+			for _, validSubject := range rule.ValidSubjects {
+				vaildCertSubject := validSubject.ToName()
+				if vaildCertSubject.ToRDNSequence().String() == subject.ToRDNSequence().String() {
+					logger.Debug("chain", zap.String("issuer", cert.Issuer.String()), zap.Bool("CA", cert.IsCA), zap.String("subject", cert.Subject.String()))
+					return nil
+				}
+			}
+			logger.Warn("invalid-subject", zap.String("issuer", cert.Issuer.String()), zap.String("subject", cert.Subject.String()), zap.Object("allowed", rule.ValidSubjects))
+			return fmt.Errorf("subject not in the list of allowed subjects for CA Subject %q: %q", rule.CASubject, subject)
+		}
+	}
```
```suggestion
	for _, cert := range chain {
		if cert.IsCA {
                     continue
		}
		subject := cert.Subject
		for _, validSubject := range rule.ValidSubjects {
			vaildCertSubject := validSubject.ToName()
			if vaildCertSubject.ToRDNSequence().String() == subject.ToRDNSequence().String() {
				logger.Debug("chain", zap.String("issuer", cert.Issuer.String()), zap.Bool("CA", cert.IsCA), zap.String("subject", cert.Subject.String()))
				return nil
			}
		}
		logger.Warn("invalid-subject", zap.String("issuer", cert.Issuer.String()), zap.String("subject", cert.Subject.String()), zap.Object("allowed", rule.ValidSubjects))
		return fmt.Errorf("subject not in the list of allowed subjects for CA Subject %q: %q", rule.CASubject, subject)	
	}
```

b1tamara replied:
@peanball a small suggestion to keep the code simple

peanball replied:
integrated. thank you!
